ADR/adr-0008-environment-provisioning.md

```diff
@@ -0,0 +1,458 @@
+- Opens the possibility to create [DeploymentTargetClaims] in advance and by that reducing the waiting time for an ephemeral environment for testing.
+- The design is similar to the design of storage in k8s, so it should look familiar to developers and users.
+- We have a larger API surface than we might ultimately need. Users won’t experience this in the UI, but API-based users will face apparent complexity. The fact that this design emulates the concepts in the k8s storage API should reduce the cognitive load this might impose on users and our engineers.
+
+## Phases of implementation
+
+A addendum about what features can be left out until later iterations.
+
+### Phase 0 (Path to MVP)
+
+- At minimum, create the CRDs and make them available, with no binder or provisioner controllers behind them.
+- Next, modify the Environment CRD and teach the gitops service how to navigate from the linked [DTC] to the [DT] in order to find the Secret that it needs for Argo.
+
+Minimal functionality looks like the BYOC case: No provisioners
+necessary - and even the binder can be omitted in this first pass.
+
+- On workspace initialization, have the NsTemplateTier:
+  - Create Namespaces with fixed -dev and -stage suffixes manually as a part of the appstudio tier. So every provisioned workspace/space has three namespaces. The main user namespace and two -dev and -stage for environments.
+  - Create serviceaccounts with rolebindings to administer the namespaces
+  - Create [DeploymentTargets] to match and reference Secrets that will be created to contains tokens for the serviceaccounts in the previous steps.
+  - Create [DeploymentTargetClaims] to match.
+  - Create [Environments] to match.
+- As long as the secret names generated by the SpaceRequests are predictable, then it should be possible to specify all of these objects up front.
+
+### Phase 1 (Service Preview)
+
+The result of this phase is an automation that binds [DT] and [DTC] and automatically provisioned
+environment on the Sandbox cluster (using the [SpaceRequest
+API](https://docs.google.com/document/d/1uqgghk1lN9dyoBLsvn5YD443TwKrsKHU6fvfUuo84Hs/edit#)) and
+creates a [DT].
+
+The following controllers should be implemented:
+
+- Binding controller.
+- Sandbox provisioner.
+- Adjusting the integration service controller to create and delete a [DTC]
+
+### Phase 2 (need more grooming and clear requirement from the PM)
+
+The result of this phase is the ability to specify parametes that are
+needed from the deployment target, such has memory, CPU, CPU
+architecture and the number of nodes.
+
+- The [DT], [DTC] an [DTCLS] would need to extended to support the new parameters.
+  - **DTCLS - allowTargetResourceExpansion** - represents whether or not the underlying provisioner allows targets to be resized. I.e., for a whole cluster provided by the "hypershift" ocm provider - can it have its number or size of nodes increased without having to delete it and create a new one.
```

Comment:
It's not important at all for early Phases. We suspect it might be useful in future phases, in the same way that it's useful to be able to scale up a PVC requested size for storageClasses that support it. However - there are a lot more kinds of parameters here for "size" than for storage. (Think limits vs requests vs node count vs node size and more..). We originally had a sketch of how that might work but it needs its own followup ADR to explore it fully. "Needs more grooming and input from PM."